janitor/utils.py

```diff
@@ -42,99 +43,117 @@ def check(varname: str, value, expected_types: list):
+        return {(key, 0): value}
+
+    return {(key, num): arr for num, arr in enumerate(value.T)}
+
+
+@_expand_grid.register(pd.arrays.PandasArray)
+def _sub_expand_grid(value, grid_index, key):  # noqa: F811
+    """
+    Expands the pandas array based on `grid_index`.
+
+    Returns a dictionary.
+    """
+
+    value = value[grid_index]
+
+    return {(key, 0): value}
 
 
 @_expand_grid.register(pd.Series)
-def _sub_expand_grid(value, grid_index):  # noqa: F811
+def _sub_expand_grid(value, grid_index, key):  # noqa: F811
     """
     Expands the Series based on `grid_index`.
 
-    Returns Series.
+    Returns a dictionary.
     """
-    if value.empty:
-        raise ValueError("""Series cannot be empty.""")
 
-    value = value.iloc[grid_index]
-    value.index = pd.RangeIndex(start=0, stop=len(value))
+    name = value.name
+    if not name:
+        name = 0
+    value = extract_array(value, extract_numpy=True)[grid_index]
 
-    return value.to_frame()
+    return {(key, name): value}
 
 
 @_expand_grid.register(pd.DataFrame)
-def _sub_expand_grid(value, grid_index):  # noqa: F811
+def _sub_expand_grid(value, grid_index, key):  # noqa: F811
     """
     Expands the DataFrame based on `grid_index`.
 
-    Returns a DataFrame.
+    Returns a dictionary.
     """
-    if value.empty:
-        raise ValueError("""DataFrame cannot be empty.""")
 
-    value = value.iloc[grid_index]
-    value.index = pd.RangeIndex(start=0, stop=len(value))
+    # use set_axis here, to prevent the column change from
+    # transmitting back to the original dataframe
     if isinstance(value.columns, pd.MultiIndex):
-        value.columns = ["_".join(map(str, ent)) for ent in value]
+        columns = ["_".join(map(str, ent)) for ent in value]
+        value = value.set_axis(columns, axis="columns")
 
-    return value
+    return {
+        (key, name): extract_array(val, extract_numpy=True)[grid_index]
+        for name, val in value.items()
+    }
 
 
 @_expand_grid.register(pd.Index)
-def _sub_expand_grid(value, grid_index):  # noqa: F811
+def _sub_expand_grid(value, grid_index, key):  # noqa: F811
     """
     Expands the Index based on `grid_index`.
 
-    Returns a DataFrame (if MultiIndex), or a Series.
+    Returns a dictionary.
     """
-    if value.empty:
-        raise ValueError("""Index cannot be empty.""")
-
-    value = value[grid_index]
 
-    return value.to_frame(index=False)
+    contents = {}
+    if isinstance(value, pd.MultiIndex):
+        num = 0
+        for n in range(value.nlevels):
+            arr = value.get_level_values(n)
+            name = arr.name
+            arr = extract_array(arr, extract_numpy=True)[grid_index]
+            if not name:
+                name = num
+                num += 1
+            contents[(key, name)] = arr
+        return contents
+    name = value.name
+    if not name:
+        name = 0
+    return {(key, name): extract_array(value, extract_numpy=True)[grid_index]}
```

Comment:
It seems to me that these two should be separate dispatch functions (one for `pd.Index`, one for `pd.MultiIndex`)

Reply:
The thing is pd.Index is the superset, so any dispatch to multiindex will go instead to the pd.index and error out. Open to suggestions on how to go about it

Reply:
Could you give me an example? I can't quite reproduce it.

```python
import pandas as pd
from functools import singledispatch

@singledispatch
def foo(x):
    raise TypeError

@foo.register(pd.MultiIndex)
def _foo(x):
    print("I'm in multiindex dispatch")

@foo.register(pd.Index)
def _foo(x):
    print("I'm in index dispatch")


m = pd.Index([1, 2, 3])
foo(m)  # dispatch to index
m = pd.MultiIndex.from_tuples([(1, 1), (2, 2), (3, 3)])
foo(m)  # dispatch to multindex
```
<img width="529" alt="image" src="https://user-images.githubusercontent.com/30731072/156874860-0b823cec-14cb-469c-b4eb-12294b8ab1f9.png">